### PR TITLE
Democracy.fast_track not allowed with zero voting period

### LINKED ISSUE #8896: democracy.fastTrack() may broke democracy pallet.
**Environment:** LOCAL

# **Detail description:**
democracy.fastTrack(proposal_hash, voting_period, delay)
Hasn't limit for voting_period and if set it as "0" created proposal will never end. And incorrect showed on polkadot{.js}:
<details>
  <summary>Image</summary>
Screenshot 2021-05-21 at 18 45 07
</details>

*Precondition:*
InstantOrigin=True

## *Scenario:*

1. Submit preimage
2. Create Propose external through Governance->Consil->Motions->Propose external
3. Voting and close
3. Create democracy.fastTrack(proposal_hash, voting_period=0, delay=0) from technical committee with proposal.
4. Voting and close

## **Actual result:**
The proposal goes to "referenda", but not finished, stay in this state forever.
democracy.referendumInfoOf(ReferendumIndex)
`{
  Ongoing: {
    end: 723,
    proposalHash: 0x36d1e29a27536522fd0bcee0d4ebacc1f33931c23319d71c1835b3ca216dfa9a,
    threshold: Simplemajority,
    delay: 0,
    tally: {
      ayes: 0,
      nays: 0,
      turnout: 0
    }
  }
}`
## **Expected result:**
Can't create a proposal with 0 voting period
or
Such proposals should finish instantly as no voting.

### PR DESCRIPTION
**Pull request by muharem**
Fixes #8896 (other possible solutions are discussed in the issue comments).

According to the issue fast track with zero voting period create referenda which never ends.
In this PR we do not allow to fast track with zero voting period, returning an error.

### CODE REVIEW

**Comment by shawntabrizi** [comment]
Need to actually look into this one, but you are saying that if you create a fast_track with voting period of 1 block then there isn't some issue?

**Comment by muharem** [comment]
> Need to actually look into this one, but you are saying that if you create a fast_track with voting period of 1 block then there isn't some issue? In this case is having 0 blocks vs 1 blocks different?

I will check this, and share what I found. 
I am assuming when a referendum expires, something is triggered, and referendum doesn't stuck in unfinished state like in issue #8896. And when voting_period 0, its already expired. 

Meanwhile can you check my comment in the issue #8896. 
There is another possible solution, but I need a bigger context for it. Question 2.
https://github.com/paritytech/substrate/issues/8896#issuecomment-1154991734

**Comment by muharem** [comment]
There are two unverified commits at the beginning.
I had to regenerate the keys.

**Comment by shawntabrizi** [comment]
Can you please add a test that someone adds a fast track with 1 block voting period, that it fails after one block, and is cleaned up correctly.

**Comment by muharem** [comment]
> Can you please add a test that someone adds a fast track with 1 block voting period, that it fails after one block, and is cleaned up correctly.

@shawntabrizi done!

File: frame/democracy/src/tests/fast_tracking.rs [hunk 1]
```diff
@@ -81,6 +85,60 @@ fn instant_referendum_works() {
...
+#[test]
+fn instant_next_block_referendum_backed() {
+	new_test_ext().execute_with(|| {
+		// arrange
+		let start_block_number = 10;
+		let majority_origin_id = 3;
+		let instant_origin_id = 6;
+		let voting_period = 1;
+		let proposal_hash = set_balance_proposal_hash_and_note(2);
+		let delay = 2; // has no affect on test
```
Comment (joepetrowski) [inline frame/democracy/src/tests/fast_tracking.rs:97]:
```suggestion
		let delay = 2; // has no effect on test
```

File: frame/democracy/src/tests/fast_tracking.rs [hunk 2]
```diff
@@ -81,6 +85,60 @@ fn instant_referendum_works() {
...
+			Ok(ReferendumStatus {
+				end: start_block_number + voting_period,
+				proposal_hash,
+				threshold: VoteThreshold::SimpleMajority,
+				delay,
+				tally: Tally { ayes: 0, nays: 0, turnout: 0 },
+			})
+		);
+
+		// referendum expected to be backed with the start of the next block
```
Comment (joepetrowski) [inline frame/democracy/src/tests/fast_tracking.rs:129]:
```suggestion
		// referendum expected to be baked with the start of the next block
```

**Review by joepetrowski — APPROVED** [review]

**Review by gilescope — APPROVED** [review]
LGTM

**Review by shawntabrizi — APPROVED** [review]

File: frame/democracy/src/tests/fast_tracking.rs (hunk 2 above)
Comment (louismerlin) [reply frame/democracy/src/tests/fast_tracking.rs]: This should actually be `backed at`, not `baked with`.

Comment (joepetrowski) [reply frame/democracy/src/tests/fast_tracking.rs]: It's referring to baking (checking if approved and scheduling dispatch) at the beginning of the expiring block. The bug this is fixing was not that a 0-length fast track had no _backing_, it's that _baking_ takes place at the start of a block, so zero-length referenda never get baked.

https://github.com/paritytech/substrate/blob/60b945caca440fdf6a7be96b7dc0db54923f039d/frame/democracy/src/lib.rs#L1783

Comment (louismerlin) [reply frame/democracy/src/tests/fast_tracking.rs]: Got it, thank you for the context :+1:

**Comment by muharem** [comment]
bot merge

**Pull request merged**